File: javascript/linked-list/reverse.js
// Given the head of a singly linked list, 
// reverse the list, and return the reversed list.

// impelemnting node 
class ListNode {
    constructor(val) {
        this.val = val
        this.next = null
    }
}

// Implementing a Linked List
class LinkedList {
    constructor(head = null) {
        this.head = head
        this.tail = head
    }

    add(node) {
        if (!this.head) {
            this.head = node
            this.tail = node
        } else {
            this.tail.next = node
            this.tail = node
        }
    }
}

let node1 = new ListNode(1)
let node2 = new ListNode(2)
let node3 = new ListNode(3)
let node4 = new ListNode(4)
let node5 = new ListNode(5)

let list = new LinkedList(node1)
list.add(node2)
list.add(node3)
list.add(node4)
list.add(node5)

function reverseList(head) {
    let prev = null
    let current = head

    while (current) {
        let next = current.next
        current.next = prev
        prev = current
        // move to the next list node
        current = next
    }

    return prev
}

// testing
console.log(reverseList(list.head)) 
// expected output: ListNode {val: 5, next: ListNode{}}

File: python/linked-list/reverse.py
# Given the head of a singly linked list, reverse the list, 
# and return the reversed list.

# Implementing a List Node
class Node:
    def __init__(self, val) -> None:
        self.val = val
        self.next = None

# Implementing a Linked List
class LinkedList:
    def __init__(self, head) -> None:
        self.head = head
        self.tail = head
    
    def add(self, node):
        if not self.head:
            self.head = node
            self.tail = node
        else:
            self.tail.next = node
            self.tail = node

node1 = Node(1)
node2 = Node(2)
node3 = Node(3)
node4 = Node(4)
node5 = Node(5)

linkedList = LinkedList(node1)
linkedList.add(node2)
linkedList.add(node3)
linkedList.add(node4)
linkedList.add(node5)

def reverseList(head: Node) -> Node:
    prev = None
    current = head
    
    while current:
        nextNode = current.next
        current.next = prev
        prev = current
        current = nextNode
    
    return prev.val

# testing
print(reverseList(linkedList.head))
# expected output: 5 (val of the head node)
